simulator.py

```diff
@@ -0,0 +1,126 @@
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        While calculating pnl,
+        we should take a look at our asset and trading cost.
+            (1) Total asset
+                We assume that our assets are the same everyday (equal to
+                self.initial_asset). To achieve this goal, we open a virtual
+                account to borrow/lend money to the investing account (this
+                account is self.cur_asset[self.base]).
+            (2) Trading cost
+                There are different functions about calculating trading cost
+                in different investing targets.
+                This method should import different modules while targets are
+                different.
+
+                In our template function, we calculate trading cost while
+                selling out the asset.
+                The trading_cost function could be built by alpha writers.
+                The API will adjust in the future, while the fuction needs
+                more data.
+        """
+        self.cost_object.reset()
+        today = self.__portfolio[-1]
+        target_data = self.data['target_data']
+        for instrument in today:
+            if target_data[instrument][-1].sell is not None:
+                self.cost_object(today, instrument, target_data[instrument][-1])
+        profit, cost = self.cost_object.result()
+        self.__result.append(
+            (today.date, profit - cost, cost, self.__portfolio[-1],
+             self.__portfolio[-1].position_np))
+
+    def run(self):
+        """
+        Carried out simulating process.
+        """
+        self.setting()
+        self.alpha_formula = self.settings.alpha(
+            self.settings.start_date, self.data, self.settings.parameters)
+        while (self.data['workdays'].get_today() is None or
+               self.data['workdays'].get_today() <= self.settings.end_date):
+            self.produce_portfolio(self.data['workdays'].get_today())
+            self.next_date()
+            if self.data['workdays'].get_today() is not None:
+                self.produce_pnl()
+            else:
+                raise ValueError("Please renew the data to calculate pnl.")
+        return self.__result
+
+    def setting(self):
+        """
+        Alpha setting.
+
+        This method is used to initialize the settings before simulating alpha.
+        We push keys into all_data, and set them into start_date.
+        """
+        self.data['workdays'].set_key(self.__key)
+        self.data['target_data'].set_key(self.__key)
+        self.data['target_data'].set_date(self.settings.start_date, self.__key)
+        for dataset in self.data['others'].values():
+            dataset.set_key(self.__key)
+            dataset.set_date(self.settings.start_date, self.__key)
+        self.data['workdays'].set_date(
+            self.settings.start_date, auth_key=self.__key)
```

Comment:
Is there any reason that this is not put right after line 119?

Reply:
In the past code, set_workday was write in this method, so put this line here.
It's legacy forget to fix.
Also, I wonder whether put all the code in this method into __init__ is a good outlook ?

Reply:
Operations here are tricky, so putting them all into `__init__` may be too messy.
I suggest to rename this method to `initialize_data`, and call this method in `__init__`.